File: src/test/java/compiler/parser/Parser.java
package compiler.parser;

import compiler.block.Block;
import compiler.tokenizer.Tokenizer;

public abstract class Parser<T extends Block> {

    /**
     * Takes a line and checks to see ifs it is for this parser by using regex.
     */
    public abstract boolean shouldParse(String line);

    /**
     * Take the superBlock and the tokenizer for the line and return a block of this parser's type.
     */
    public abstract T parse(Block superBlock, Tokenizer tokenizer);
}

File: src/main/scala/compiler/parser/Parser.scala
package compiler.parser

import compiler.block.Block
import compiler.tokenizer.Tokenizer

abstract class Parser[T <: Block] {
  /**
    * Takes a line and checks to see ifs it is for this parser by using regex.
    */
  def shouldParse(line: String): Boolean

  /**
    * Take the superBlock and the tokenizer for the line and return a block of this parser's type.
    */
  def parse(superBlock: Block, tokenizer: Tokenizer): Block
}